File: src/ca/rightsomegoodgames/mayacharm/run/debug/MayaCharmDebugProcess.java
package ca.rightsomegoodgames.mayacharm.run.debug;

import ca.rightsomegoodgames.mayacharm.mayacomms.MayaCommInterface;
import ca.rightsomegoodgames.mayacharm.settings.MCSettingsProvider;
import com.intellij.execution.process.ProcessHandler;
import com.intellij.execution.ui.ExecutionConsole;
import com.intellij.xdebugger.XDebugSession;
import com.jetbrains.python.debugger.PyRemoteDebugProcess;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.net.ServerSocket;

public class MayaCharmDebugProcess extends PyRemoteDebugProcess {
    final private ServerSocket socket;
    final private MayaCharmDebugConfig config;
    final private MCSettingsProvider settings;
    final private MayaCommInterface mayaCommInterface;
    private boolean canConnect;

    public MayaCharmDebugProcess(@NotNull XDebugSession xDebugSession, @NotNull ServerSocket serverSocket, @NotNull ExecutionConsole executionConsole, @Nullable ProcessHandler processHandler, @Nullable String s) {
        super(xDebugSession, serverSocket, executionConsole, processHandler, s);
        socket = serverSocket;
        config = (MayaCharmDebugConfig) xDebugSession.getRunProfile();
        settings = MCSettingsProvider.getInstance(getProject());
        mayaCommInterface = new MayaCommInterface(settings.getHost(), settings.getPort());
    }

    @Override
    public void sessionInitialized() {
        canConnect = true;
        super.sessionInitialized();
    }

    @Override
    protected void beforeConnect() {
        if (canConnect) {
            mayaCommInterface.pyDevSetup();
            mayaCommInterface.setTrace(socket.getLocalPort(), config.isSuspendAfterConnect(), config.isRedirectOutput());
        }
        super.beforeConnect();
    }

    @Override
    protected void afterConnect() {
        canConnect = false;
        super.afterConnect();
        if (config.getUseCode()) {
            mayaCommInterface.sendCodeToMaya(config.getScriptCodeText());
        }
        else {
            mayaCommInterface.sendFileToMaya(config.getScriptFilePath());
        }
    }

    @Override
    public void stop() {
        canConnect = false;
        mayaCommInterface.stopTrace();
        super.stop();
    }

    @Override
    protected void disconnect() {
        canConnect = false;
        super.disconnect();
    }
}

File: src/main/kotlin/run/MayaCharmDebugProcess.kt
package run

import MayaBundle as Loc
import mayacomms.MayaCommandInterface
import settings.ApplicationSettings
import com.intellij.execution.process.ProcessHandler
import com.intellij.execution.ui.ConsoleViewContentType
import com.intellij.execution.ui.ExecutionConsole
import com.intellij.xdebugger.XDebugSession
import com.jetbrains.python.debugger.PyDebugProcess
import resources.PythonStrings
import java.net.ServerSocket

class MayaCharmDebugProcess(
    session: XDebugSession,
    serverSocket: ServerSocket,
    executionConsole: ExecutionConsole,
    processHandler: ProcessHandler?,
    private val runConfig: MayaCharmRunConfiguration?,
    private val pid: Int
) : PyDebugProcess(session, serverSocket, executionConsole, processHandler, false) {

    private val mayaCommand = runConfig?.mayaSdkPath?.let {
        val port = ApplicationSettings.INSTANCE.mayaSdkMapping[it]?.port ?: return@let null
        MayaCommandInterface(port)
    }

    override fun getConnectionMessage(): String {
        return Loc.message("mayacharm.debugproc.ConnectionMessage", pid.toString())
    }

    override fun getConnectionTitle(): String {
        return Loc.message("mayacharm.debugproc.ConnectionTitle")
    }

    override fun afterConnect() {
        if (!isConnected) {
            printToConsole(Loc.message("mayacharm.debugproc.FailedToConnect"), ConsoleViewContentType.SYSTEM_OUTPUT)
            return
        }

        runConfig ?: return
        mayaCommand ?: return

        when (runConfig.executionType) {
            ExecutionType.FILE -> mayaCommand.sendFileToMaya(runConfig.scriptFilePath)
            ExecutionType.CODE -> mayaCommand.sendCodeToMaya(runConfig.scriptCodeText)
        }
    }

    override fun disconnect() {
        super.disconnect()
        mayaCommand?.sendCodeToMaya(PythonStrings.STOPTRACE.message)
    }
}
